apps/emqx_bridge_pulsar/src/emqx_bridge_pulsar_impl_producer.erl

```diff
@@ -0,0 +1,399 @@
+        bridge_name := BridgeName,
+        servers := Servers0,
+        ssl := SSL
+    } = Config,
+    Servers = format_servers(Servers0),
+    ClientId = make_client_id(InstanceId, BridgeName),
+    SSLOpts = emqx_tls_lib:to_client_opts(SSL),
+    ClientOpts = #{
+        ssl_opts => SSLOpts,
+        conn_opts => conn_opts(Config)
+    },
+    case pulsar:ensure_supervised_client(ClientId, Servers, ClientOpts) of
+        {ok, _Pid} ->
+            ?SLOG(info, #{
+                msg => "pulsar_client_started",
+                instance_id => InstanceId,
+                pulsar_hosts => Servers
+            });
+        {error, Reason} ->
+            ?SLOG(error, #{
+                msg => "failed_to_start_pulsar_client",
+                instance_id => InstanceId,
+                pulsar_hosts => Servers,
+                reason => Reason
+            }),
+            throw(failed_to_start_pulsar_client)
+    end,
+    start_producer(Config, InstanceId, ClientId, ClientOpts).
+
+-spec on_stop(manager_id(), state()) -> ok.
+on_stop(_InstanceId, State) ->
+    #{
+        pulsar_client_id := ClientId,
+        producers := Producers
+    } = State,
+    stop_producers(ClientId, Producers),
+    stop_client(ClientId),
+    ?tp(pulsar_bridge_stopped, #{instance_id => _InstanceId}),
+    ok.
+
+-spec on_get_status(manager_id(), state()) -> connected | disconnected.
+on_get_status(_InstanceId, State) ->
+    #{
+        pulsar_client_id := ClientId,
+        producers := Producers
+    } = State,
+    case pulsar_client_sup:find_client(ClientId) of
+        {ok, Pid} ->
+            try pulsar_client:get_status(Pid) of
+                true ->
+                    get_producer_status(Producers);
+                false ->
+                    disconnected
+            catch
+                error:timeout ->
+                    disconnected;
+                exit:{noproc, _} ->
+                    disconnected
+            end;
+        {error, _} ->
+            disconnected
+    end.
+
+-spec on_query(manager_id(), {send_message, map()}, state()) -> ok | {error, timeout}.
+on_query(_InstanceId, {send_message, Message}, State) ->
+    #{
+        producers := Producers,
+        sync_timeout := SyncTimeout,
+        message_template := MessageTemplate
+    } = State,
+    PulsarMessage = render_message(Message, MessageTemplate),
+    try
+        pulsar:send_sync(Producers, [PulsarMessage], SyncTimeout)
+    catch
+        error:timeout ->
+            {error, timeout}
+    end.
+
+-spec on_query_async(
+    manager_id(), {send_message, map()}, {ReplyFun :: function(), Args :: list()}, state()
+) ->
+    {ok, pid()}.
+on_query_async(_InstanceId, {send_message, Message}, AsyncReplyFn, State) ->
+    #{
+        producers := Producers,
+        message_template := MessageTemplate
+    } = State,
+    PulsarMessage = render_message(Message, MessageTemplate),
+    pulsar:send(Producers, [PulsarMessage], #{callback_fn => AsyncReplyFn}).
```

Comment:
I wonder what is the contract here? Is it by accident that it is compatible wrt number and type of arguments passed into the callback when called?

Reply:
Sorry, which contract?  For `callback_fn`?  Didn't quite understand the question.

`AsyncReplyFn :: {function(), [term()]}`, and that's the expected type for `callback_fn`.

Reply:
Let me try to rephrase. It's the framework's code that passes `callback_fn` to `emqx_bridge_pulsar_impl_producer:on_query_async` - there's a contract (what you wrote above) between the two to ensure it's being used the right way. But `pulsar:send` is out of the scope. It just happens to adhere to the same idea.

But I get it now, it's our own library. So of course it follows the same pattern. 

Eg, the async producer in brod looks different, IIRC. I think it's just sth like `CallbackFn(Result)`. So if you had a bridge with a client like that, you'd have to wrap things somehow in your bridge implementation. 

For someone who doesn't know about what you explained it just looks a bit like a mix of  "good luck and magic".

Reply:
Yes, if `pulsar:send` expected other formats, we would need to transform or make a function adhering to its expectations before passing it on, indeed.